src/AccessLogs.sol

```diff
@@ -0,0 +1,190 @@
+    // Read methods
+    //
+    function readRegion(
+        AccessLogs.Context memory a,
+        Memory.Region memory region
+    ) internal pure returns (bytes32) {
+        bytes32 drive = a.buffer.toBytes32(a.pointer);
+        a.pointer += 32;
+        (bytes32 rootHash, uint8 siblingCount) =
+            getRoot(region, drive, a.buffer, a.pointer);
+        a.pointer += uint128(siblingCount) * 32;
+
+        require(a.currentRootHash == rootHash, "Read region root doesn't match");
+
+        return drive;
+    }
+
+    function readLeaf(AccessLogs.Context memory a, Memory.Stride readStride)
+        internal
+        pure
+        returns (bytes32)
+    {
+        Memory.Region memory r =
+            Memory.regionFromStride(readStride, Memory.alignedSizeFromLog2(0));
+        return readRegion(a, r);
+    }
+
+    function readWord(
+        AccessLogs.Context memory a,
+        Memory.PhysicalAddress readAddress
+    ) internal pure returns (uint64) {
+        uint64 val =
+            uint64SwapEndian(uint64(bytes8(a.buffer.toBytes32(a.pointer))));
+        a.pointer += 8;
+        bytes32 valHash = keccak256(abi.encodePacked(uint64SwapEndian(val)));
+        bytes32 expectedValHash =
+            readLeaf(a, Memory.strideFromWordAddress(readAddress));
+
+        require(valHash == expectedValHash, "Read value doesn't match");
+        return val;
+    }
+
+    //
+    // Write methods
+    //
+    function writeRegion(
+        AccessLogs.Context memory a,
+        Memory.Region memory region,
+        bytes32 newHash
+    ) internal pure {
+        bytes32 oldDrive = a.buffer.toBytes32(a.pointer);
+        a.pointer += 32;
+        (bytes32 rootHash, uint8 siblingCount) =
+            getRoot(region, oldDrive, a.buffer, a.pointer);
+
+        require(
+            a.currentRootHash == rootHash, "Write region root doesn't match"
+        );
+
+        (bytes32 newRootHash,) = getRoot(region, newHash, a.buffer, a.pointer);
+        a.pointer += uint128(siblingCount) * 32;
+
+        a.currentRootHash = newRootHash;
+    }
+
+    function writeLeaf(
+        AccessLogs.Context memory a,
+        Memory.Stride writeStride,
+        bytes32 newHash
+    ) internal pure {
+        Memory.Region memory r =
+            Memory.regionFromStride(writeStride, Memory.alignedSizeFromLog2(0));
+        writeRegion(a, r, newHash);
+    }
+
+    function writeWord(
+        AccessLogs.Context memory a,
+        Memory.PhysicalAddress writeAddress,
+        uint64 newValue
+    ) internal pure {
+        writeLeaf(
+            a,
+            Memory.strideFromWordAddress(writeAddress),
+            keccak256(abi.encodePacked(uint64SwapEndian(newValue)))
+        );
+    }
+
+    uint8 constant LOG2RANGE = 61;
+
+    function isEven(uint64 x) private pure returns (bool) {
+        return x % 2 == 0;
+    }
+
+    function getRoot(
+        Memory.Region memory region,
+        bytes32 drive,
+        bytes memory siblings,
+        uint128 offset
```

Comment:
These two arguments are the new `Buffer` type. If we create it, do you think maybe this could be a method of the `Buffer` type? I don't know if it makes sense.

Reply:
I think we'll need two functions, one to be `getRootWithoutPointerAdvanced`, the other one to be `getRootWithPointerAdvanced`. Because in `writeRegion` method, we have to `getRoot` twice, the first one being validation, the second one to calculate new hash. And the pointer cannot be advanced in between the calls.

Reply:
Yeah, hadn't thought of that! I like it.